### ISSUE TITLE
`Container { exitCode }` doesn't work if the command fails

### ISSUE THREAD

**Issue by vito** [dagger/dagger#3192]
We currently expose this API:

```graphql
type Container {
  exitCode: Int
}
```

...but the only possible value returned here is `null` or `0`. If the command exits nonzero Buildkit will error instead.

We could have the shim mask the exit status, but we would probably only want to do that if the user is requesting `exitCode`, but then that changes the underlying LLB. 🤔 

The same is true for `stdout` and `stderr` that share the same underlying code.

**Comment by sipsma** [comment]
Yeah this another one that's harder than it looks. I completely agree we can't have the shim exit 0 as that'll cache the failure.

My immediate thought is to do something like:
1. Have the shim check the error code and then write it to a mount (similar to the mounts used for stdout/stderr today).
2. Have the shim then exit with that error too
3. Use this mechanism to grab the FS changes after the error occurs: https://github.com/sipsma/buildkit/blob/61307b8f2e73d681721b51856c10cb8606e5ea70/client/build_test.go#L1284-L1284
4. Use that to return the exit code

I actually think we might end up needing to do something similar to fix this issue too: https://github.com/dagger/dagger/issues/3025

Take this with a grain of salt, I haven't used that whole `SolveError` mechanism before and I also would love it if there was something simpler; this is just what comes to mind right away.

**Comment by sipsma** [comment]
Another implication of the above is that the execop failure is no longer a graphql error like it is today. I think that's fine? But it is different than today.

**Comment by aluzzardi** [comment]
I was thinking eventually `exec` would take some options (e.g. `continueOnFailure`, in which case `exitCode` will be non-zero ... maybe at some point some retry options?)

Main problem is what @sipsma pointed out -- if the shim exits with 0, then buildkit will cache the failure and not retry (e.g. `apk add curl` would end up "cached", even if there was a transient network issue, and will not attempt to retry)

**Comment by GiedriusS** [comment]
This is probably the biggest blocker for me ATM. I have even resorted to doing something like this as a workaround:

```go
lint = lint.Exec(dagger.ContainerExecOpts{
		Args: []string{"/bin/sh", "-c", `/go/bin/golangci-lint run -v && printf OK > /ok || printf FAIL > /ok`},
	})

f := lint.File("/ok")
content, err := f.Contents(ctx)
if err != nil {
return err
}
if content != "OK" {
return fmt.Errorf("linting failed (got %s)", content)
}
```

**Comment by aluzzardi** [comment]
> This is probably the biggest blocker for me ATM. I have even resorted to doing something like this as a workaround:

@GiedriusS As a workaround, what if you just let it fail on its own?

```go
_, err := lint.Exec(dagger.ContainerExecOpts{
  Args: []string{"/bin/sh", "-c", `/go/bin/golangci-lint run -v`},
}).ExitCode()
if err != nil {
  return err
}
```

In any case, logs are printed to the console (if using `dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))`), so you can get away without capturing stderr in code.

**Comment by aluzzardi** [comment]
Bumping this is priority

Here's a quick and dirty proposal:

- Shim communicates with core
  - Perhaps we mount another gRPC service in containers. Or it could be a dumb HTTP POST, whatever
  - Shim reports back stdout, stderr, exitCode **in addition to** storing to /meta
  - We probably need an identifier to come back via gRPC so we know which container is talking to us. Could potentially pack it in HTTP_PROXY as @sipsma mentioned (as it won't bust the cache)
- We add some new control options to WithExec
  - e.g. `abortOnFailure`?
  - In which case, when an exec operation fails, we don't fail the entire query and instead use stdout/stderr/exitCode from the shim

Shortcomings:

- This is not re-entrant: "streamed" shim metadata lives in memory. I think this could have a bunch of side effects as it's a very different model than what we have with Container ID
- `abortOnFailure` is stranger than it looks with nesting
  - `withExec("exit 1").WithExec("true").ExitCode()` --> if we continue on failure, what is exit code even supposed to be?

After typing this, I don't think this is a viable option.

**Comment by sipsma** [comment]
@aluzzardi 

> Perhaps we mount another gRPC service in containers. Or it could be a dumb HTTP POST, whatever

I don't think we'd need this. If we allow an exec to fail without it being an actual error returned to the client (via `abortOnFailure` or similar) then we can just store all the metadata in /meta and have the shim exit 0. The real exit code (plus stdout/err) will be easily accessible still.

The other thing is I think we can improve most of the questions we get around this by just including the output of the exec in the graphql error when it fails. That's simpler than trying to figure out how to allow for failure not result in a graphql error at all: https://github.com/dagger/dagger/issues/3025

**Comment by sagikazarmark** [comment]
I have the same problem with Stdout and Stderr. I worked around the problem with a DEBUG env var that sets `dagger.WithLogOutput(os.Stderr)`, but for newcomers this would definitely be super confusing (eg. linter fails, but no actual error is shown).

**Comment by shykes** [comment]
Given the urgency of the problem, and the complexity of solving it at the `withExec` layer, perhaps we should consider a stopgap?

What if we added a slightly higher-level `withShellCommand` (built on `withExec` under the hood), that executes the specified shell command... But with the shell executed in such a way that a failed execution is caught (it would just generate a wrapper shell script).

**Comment by aluzzardi** [comment]
@sagikazarmark: @sipsma made a change a while ago to bubble up stdout/stderr in the returned `err` -- is that enough as a stopgap?

e.g.

```go
_, err := WithExec("fail").ExitCode()
if err != nil {
  fmt.Printf("%v\n", err) // this should include stdout
}
```

**Comment by sagikazarmark** [comment]
Works for me.

The ideal solution for me would be some helper function that returns the exit code AND the contents of stdout AND stderr. I just want to channel those from the container right back to my shell.

**Comment by mircubed** [comment]
@aluzzardi We added the "stuck" label since we need more clarity on the stopgap options. Can you document what the options are, so we can assign someone to work on this issue? 

Please @ me, so I can ensure someone is assigned as soon as you provide the options. Thanks!

**Comment by shykes** [comment]
Quoting [`@bessey` on Discord](https://discord.com/channels/707636530424053791/1080527877688275015/1080540679211593819):

> The workaround appears to omit output
> 
> If I look at the output that comes from Config log_output i.e.
> dagger.Connection(dagger.Config(log_output=sys.stderr))
> 
> 
> It contains information that 
> try:
>     exit_code = await process.exit_code()
> except Exception:
>     print(sys.exception())
> 
> 
> misses. It looks to me like printing the exit code exception only includes STDERR, not STDOUT

**Comment by vito** [comment]
Dropping a note from a conversation had with @sipsma in Discord: this should be easier to fix now that we have the plumbing for collecting the stdout/stderr/exit status that we added for better error messages.

The safest option is to keep `exitCode`/`stdout`/`stderr` fields erroring on nonzero exit by default and add a `allowFailure` (or something, please bikeshed) param for having them return the value instead of erroring.

**Comment by alrs** [comment]
I'm just trying Dagger for the first time, and it looks broken on my first pass.

`Container.Exec()` (or WithExec()`) returns an error if I execute `/bin/false`. This shouldn't be an error: dagger was able to execute the program. `Container.ExitCode()` exists to check whether it exited as `0` or something else.

**Comment by helderco** [comment]
> The safest option is to keep `exitCode`/`stdout`/`stderr` fields erroring on nonzero exit by default and add a `allowFailure` (or something, please bikeshed) param for having them return the value instead of erroring.

I understand we have to be careful not to break users' pipelines, but I feel it's too soon to be stuck with this bug in our history. If there's other breaking changes we need to make, we can try to put this in with those in the same release, but still fix it.

Using `allowFailure` (or something 😅) sounds reasonable, but for the opposite. I.e., still return the exit code by default, having to set that flag to `True` to raise an error.

### Pipeline synchronization

I'd also like us to unblock the "pipeline synchronization" issue:
- https://github.com/dagger/dagger/issues/4205

A lot of users are using `ExitCode` to execute their pipelines, not caring about the actual exit code besides success vs failure. By relying more on a "sync" feature, we'd also hopefully see lower usage for `ExitCode`, and the number of breaking pipelines because of that fix.

### SDK helpers

Another way to mitigate the breaking change would be to implement that particular pattern as an SDK helper that can be imported and used directly. If users depended on the helper in the meantime, that helper's implementation would be able to be adjusted in the same PR that fixes this bug, thereby not breaking users. 

Right now, the best workaround is for everyone to implement such function and use it globally, thus having to fix it in only one place.

**Comment by helderco** [comment]
Just referencing another confused user in:
- https://github.com/dagger/dagger/issues/4706#issuecomment-1498009538

Update: I'm currently writing a proposal to fix this in a backwards compatible way.

**Comment by helderco** [comment]
As I mentioned before, it seems to me that even though this has been flagged as a priority, it hasn't moved forward because we don't want to break users that have come to rely on the current behavior.

Here's my proposal to fix that in a new call, and deprecate `exitCode`:
- https://github.com/dagger/dagger/issues/4921

I also recognize that this was marked as stuck by @mircubed in https://github.com/dagger/dagger/issues/3192#issuecomment-1433752515, pending a summary on what the solution and current stop gaps are. The above proposal suggests an answer to that.

However, I'm assuming in the proposal that we can solve this based on:

> Dropping a note from a conversation had with @sipsma in Discord: this should be easier to fix now that we have the plumbing for collecting the stdout/stderr/exit status that we added for better error messages.

\- _Originally posted by @vito in https://github.com/dagger/dagger/issues/3192#issuecomment-1459212866_

Especially in a new encapsulating field, while making the current ones deprecated. However, I'm not entirely sure if @vito and @sipsma have considered the effect on cache that was raised by @aluzzardi here.

Can you guys confirm we can technically fix this, with @aluzzardi's concerns addressed?

**Comment by itaischwartz** [comment]
For me, this is a blocker for using Dagger in a local development environment.

**Comment by sipsma** [comment]
> Especially in a new encapsulating field, while making the current ones deprecated. However, I'm not entirely sure if @vito and @sipsma have considered the effect on cache that was raised by @aluzzardi here.

Yeah we can implement the proposal you linked to with the desired caching behavior.

That being said, there technically are two options for what the desired caching behavior is:
1. Same behavior as today: non-zero exit code means the result will not be cached.
2. New behavior: if you use this new api and exit non-zero, we will still cache the result

I strongly prefer `1`, maybe it's obvious, but seemed worth mentioning. I can technically imagine some very specific scenarios in which users could want caching of non-zero exits, but that's likely obscure and could also become a separate flag to toggle that behavior on/off at a later date if needed.

---

To implement `1`, we can re-use [the mechanism here](https://github.com/dagger/dagger/blob/3e5992dbc114789bd6bab77d4bc8d417f9350f24/core/gateway.go#L122), which is an obscure buildkit feature that lets you take an ExecOp which failed and spin up a new gateway container that has all the filesystem modifications made by the exec up to the point of failure. That currently integrates with our [shim here](https://github.com/dagger/dagger/blob/3e5992dbc114789bd6bab77d4bc8d417f9350f24/cmd/shim/main.go#L190-L210) to enable grabbing stdout/stderr.

We are actually already writing the exit code to the same mount next to stdout/stderr [here](https://github.com/dagger/dagger/blob/3e5992dbc114789bd6bab77d4bc8d417f9350f24/cmd/shim/main.go#L264-L266), so we just need to add support for grabbing that value after the failure and plumbing it through to the new API you proposed @helderco. We probably need to tweak some things so that we can return actual values instead of just annotating the error message like we do today, but it shouldn't require anything totally.

**Comment by helderco** [comment]
Thanks for the feedback @vito (Discord) and @sipsma. I'm clear on the technical feasibility, but I'm not entirely sure on caching yet.

- **`apk install`:** A failure in this kind of command could be due to a network issue, but also referencing a package that doesn't exist. You don't want to cache on the former but it's no use retrying on the latter.
- **`go test`:** Or lint, etc... most common issue in these cases is that the source code under test needs fixing. In this case I'd want to cache. The run time can easily be long but the result will be the same so it's wasteful. On the other hand there could also be a bigger issue at hand and debugging it could be harder.

It doesn't seem like it's possible to tell the difference in a smart way here, so what feels safest is to not cache by default, but add a flag to `withExec` to allow caching on failure. That would enable tuning performance for the cases where it makes sense, but it would need to be explicit (thinking about verbosity in the common case of running tests).

After writing this I realize that's basically what you said for `1`, @sipsma, so I agree. Just had to lay it out. 😁

**Comment by vito** [comment]
+1 to not caching failures by default; seems like the most ergonomic choice when it comes to network flakes etc. In an ideal world we would be able to cache `go test` failures but I think that'd be an uphill battle with folks who want to just re-run on a test flake. (They should "just" fix the tests, but...)

One use case I can think of for caching failures is if you're running a command that you _expect_ to fail and you're using it almost like a conditional. For example, you might run `ldd ./foo` to make sure that you've built a static binary. In that case failure is success! I do this in Bass [here](https://github.com/vito/bass/blob/e39afdfb0d3287797111267abb5ceaa4eedce63b/bass/bass.bass#L113-L116), the downside being it shows a scary `ldd` "failure" in the progress output.

**Comment by helderco** [comment]
The proposal to create a new encapsulating field isn’t perfect (since I haven’t found a really good DX for explicitly selecting which leaf fields you want), but I think I have convincing arguments for why it’s worth it (I’ll update the proposals to add clarity on some things).

- https://github.com/dagger/dagger/issues/4921
- Based on https://github.com/dagger/dagger/issues/4920

Since most users have come to depend on the current behavior, it seems to me that the **higher risk** is in the **breaking change** by fixing the current `exitCode`, `stdout` and `stderr` fields.

Unless there’s consensus that we should fix them anyway, this issue will remain *stuck* for a while longer until the design for the new pattern is approved.

**Comment by helderco** [comment]
> How does our today's discussion about your R&D work on exit codes affect this @helderco?

I’ve withdrawn the https://github.com/dagger/dagger/issues/4921 proposal, suggesting now a different direction:
- https://github.com/dagger/dagger/issues/5124

That issue explains everything.

**Comment by gerhard** [comment]
If you think @helderco that we should close this issue in favour of https://github.com/dagger/dagger/issues/5124, go for it! cc @mircubed

**Comment by helderco** [comment]
If I get agreement for [deprecating exitCode](https://github.com/dagger/dagger/issues/5124), this issue can be closed.

**Comment by helderco** [comment]
Closing in favor of:
- #5124
